lib/instrumentation/modules/cassandra-driver.js

```diff
@@ -0,0 +1,160 @@
+'use strict'
+
+const semver = require('semver')
+const sqlSummary = require('sql-summary')
+
+const shimmer = require('../shimmer')
+
+module.exports = function (cassandra, agent, version, enabled) {
+  if (!enabled) return cassandra
+  if (!semver.satisfies(version, '^3.0.0')) {
+    agent.logger.debug('cassandra-driver version %s not supported - aborting...', version)
+    return cassandra
+  }
+
+  if (cassandra.Client) {
+    shimmer.wrap(cassandra.Client.prototype, 'connect', wrapConnect)
+    shimmer.wrap(cassandra.Client.prototype, 'execute', wrapExecute)
+    shimmer.wrap(cassandra.Client.prototype, 'eachRow', wrapEachRow)
+    shimmer.wrap(cassandra.Client.prototype, 'batch', wrapBatch)
+  }
+
+  return cassandra
+
+  function wrapConnect (original) {
+    return function wrappedConnect (callback) {
+      const span = agent.buildSpan()
+      if (!span) {
+        return original.apply(this, arguments)
+      }
+
+      span.type = 'db.cassandra.connect'
+      span.start()
+
+      function resolve () {
+        span.end()
+      }
+
+      // Wrap the callback
+      const ret = original.call(this, wrapCallback(callback))
+
+      if (typeof callback !== 'function') {
```

Comment:
It might be safest if we checked if `ret.then` was a function as well. Just in case the API changes. In that case we can't end the span though, so we should consider what to do instead.

The best solution is probably to simply bail in that case and abort the span. I don't think we have a way to do that at the moment, but it should be too hard to add a `span.[cancel|abort|destroy]` function. That function needs to either remove the span from the `transaction._builtSpans` array or update the logic that loops over this array when the transaction ends to ensure that it's ignored instead of being truncated.

Reply:
I guess we have the same issue with not ending the span in these edge cases with all the other instrumentations we're doing. So maybe it's out of scope of this instrumentation and we should simply make another issue to improve this.

Reply:
fixed.